benchmarks/full_report/vbench/runner.py

```diff
@@ -0,0 +1,129 @@
+#  Copyright (c) 2024. Lance Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""
+This file contains the runner for the benchmarking framework.
+"""
+
+from typing import Any, Dict
+
+import attrs
+import logging
+import functools
+import numpy as np
+import time
+from tqdm.auto import tqdm
+from concurrent.futures import ThreadPoolExecutor, as_completed, wait
+
+from .experiment import Benchmark
+
+
+@attrs.define
+class RecallStats:
+    dataset_name: str
+    test_name: str
+    index_params: Dict[str, Any]
+    query_params: Dict[str, Any]
+
+    recall: float
+
+    latency_mean: float
+    latency_std: float
+
+    throughput: float
+
+
+@attrs.define
+class TestRunner:
+    benchmark: Benchmark = attrs.field(kw_only=True)
+    test_name: str = attrs.field(kw_only=True)
+    query_params: Dict[str, Any] = attrs.field(kw_only=True)
+
+    num_threads: int = attrs.field(default=1, kw_only=True)
+
+    # not actually used, just for returning this in the stats
+    index_params: Dict[str, Any] = attrs.field(kw_only=True)
+
+    def run(self) -> RecallStats:
+        logging.info("running test %s on %s", self.test_name, self.benchmark.name)
+
+        test_case = self.benchmark.test_cases[self.test_name]
+
+        k_gt = test_case.ground_truth.shape[1]
+
+        count = 0
+        total_hits = 0
+
+        latency_sum = 0
+        latency_squared_sum = 0
+
+        def _test(q, gt):
+            nonlocal latency_sum, latency_squared_sum
+            start = time.perf_counter()
+            res = self.benchmark.data.to_table(
+                nearest={"q": q, **self.query_params},
+                with_row_id=True,
+                columns=[],
+            )["_rowid"].to_numpy()
+            end = time.perf_counter()
+
+            elapsed = end - start
+            latency_sum += elapsed
+            latency_squared_sum += elapsed**2
+            return len(np.intersect1d(res, gt))
+
+        with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
+            futs = []
+
+            # warmup
+            for _ in range(100):
+                futs.append(
+                    executor.submit(
+                        functools.partial(
+                            _test,
+                            np.random.standard_normal(test_case.query.shape[1]),
+                            list(range(k_gt)),
```

Comment:
I'm not sure I get this part. Is the dataset sorted by the ground truth distance to a known query vector?

Reply:
the gt and query here are just dummies meant for warming up the threads and dataset. the result is discarded later

Reply:
Okay that makes more sense.